### ISSUE TITLE
Benefit of LambdaExpression.Compile(bool) & Expression<TDelegate>.Compile(bool) unclear.

### ISSUE THREAD

**Issue by JonHanna** [dotnet/docs#3866]
[LambdaExpression.Compile(bool)](https://msdn.microsoft.com/en-us/library/dd268054(v=vs.110).aspx) and [Expression&lt;TDelegate>.Compile(bool)](https://msdn.microsoft.com/en-us/library/dd267628(v=vs.110).aspx) both say:

> **true** to indicate that the expression should be compiled to an interpreted form, if it's available; otherwise, **false**.

There's no indication that it might the "if it's available" may be the other way around, with interpretation available but compilation not, in which case, being a preference, it happens the other way around with interpretation used even if compilation is asked for.

There's no suggestion as to why a user might possibly prefer one over the other.

The main reasons one might that I can think of is performance. In general compilation (**false**) takes longer to compile but results in a delegate that is faster to run than interpretation (**true**) and the combination of the two can mean that interpretation is **sometimes** faster for the full compile-and-run, especially from one-off expressions that don't offer the possibility of caching. Contra that, interpretation has some shortcomings (not yet well-documented either) aside from the swings-and-roundabouts performance differences.

**Comment by mairaw** [comment]
Thanks @JonHanna. @VSadov, @OmarTawfik can one of you help with this one?

**Comment by JonHanna** [comment]
@bartdesmet proposed the overload and @antiufo had said that they'd data supporting the performance advantages so may have useful things to say.

**Comment by VSadov** [comment]
@JonHanna  is right 
- Either interpreter or compiler support may not be present. In such case what is available will be used. -
- When both are available, the default behavior is to prefer compiler, but this fag coudl be used to override that.

Why user might want to force this one way or another:
- testing. Primary target supports only interpreting (like AOT), but development is done on the runtime that supports compiling. To not accidentally lock into aspects of compilation, user might want to force interpreting.
- consistency between environments when both compiler and interpreter are available. Both have limitations and customer might want to deal with just one set of liitations.
- at least in theory, depending on the shape of the code, there could be scenarios where interpreting is faster. For example when dynamic code needs to run only once. Even though compiled code could be significantly faster it would not have an opportunity to amortize the costs of JIT-compiling the method over multiple invocations. It is strongly recommended to confirm such findings by actually profiling/testing the code.

**Comment by antiufo** [comment]
@JonHanna Tomorrow I'll run some performance tests, in the meantime, these are some examples of scenarios where interpretation made sense:
* DbLinq: An `Expression.Evaluate` extension method is often used all over the codebase, mostly to handle constants/immediate scalar values
* DbLinq: Creating a `RowCreator `delegate that converts `IDataRecord `to `TRow`, when the `RowCreator `cannot be cached for future queries
* Shaman: Web interface that makes it possible to compose an (almost) arbitrary LINQ query -> syntax tree is converted into an `Expression `-> the expression is evaluated.

Basically, the interpreter makes a lot of sense if you need to evaluate many single-use, dynamically-generated `Expression`s that don't perform CPU-intensive operations (for example, they simply compose an `IQueryable`, or retrieve some values or call some methods)

[Forked version of DbLinq with the interpreter](https://github.com/antiufo/mono/commits/dblinq-customtypes)
